### PR TITLE
[Core][Test] Add ubsan support for C++ tests

### PR DESCRIPTION
**Pull request by scv119**
Run C++ test in ubsan as well.

### CODE REVIEW

**Comment by scv119** [comment]
looks TSAN/UBSAN caught some bugs. we need to see if we can fix them right now, or spread the work and enable sanitized tests gradually.

File: src/ray/thirdparty/sha256.c [hunk 1]
```diff
@@ -17,6 +17,19 @@
...
+// - __supress_ubsan__("undefined")
+// - __supress_ubsan__("signed-integer-overflow")
+// adaped from
+// https://github.com/google/flatbuffers/blob/master/include/flatbuffers/base.h
+#if defined(__clang__)
+#define __supress_ubsan__(type) __attribute__((no_sanitize(type)))
+#elif defined(__GNUC__) && (__GNUC__ * 100 + __GNUC_MINOR__ >= 409)
+#define __supress_ubsan__(type) __attribute__((no_sanitize_undefined))
+#else
+#define __supress_ubsan__(type)
```
Comment (ericl) [inline src/ray/thirdparty/sha256.c:30]:
```suggestion
#define __suppress_ubsan__(type)
```

**Review by rkooo567 — APPROVED** [review]
LGTM. We probably also want to run some python tests? Without that, things like node_manager.cc / core_worker.cc won't be sanitized I think

File: .buildkite/pipeline.yml [hunk 2]
```diff
@@ -130,6 +130,15 @@
       --test_tag_filters=-flaky
       -- //:all -//:core_worker_test
 
+- label: ":cpp: Tests (UBSAN)"
```
Comment (mwtian) [inline .buildkite/pipeline.yml:133]: If running ubsan separately has too much (testing / maintenance) overhead, we can enable ubsan together with asan too.

**Review by mwtian — APPROVED** [review]

**Pull request merged**